PluginsAndFeatures/azure-toolkit-for-intellij/src/main/java/com/microsoft/azure/toolkit/intellij/function/runner/core/FunctionCliResolver.java

```diff
@@ -5,73 +5,151 @@
+import com.microsoft.azure.toolkit.lib.common.cache.Cacheable;
+import com.microsoft.azure.toolkit.lib.common.operation.AzureOperation;
+import com.microsoft.azure.toolkit.lib.common.utils.CommandUtils;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.lang3.StringUtils;
 
+import javax.annotation.Nullable;
 import java.io.File;
 import java.io.IOException;
 import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
 import java.util.List;
+import java.util.Optional;
+import java.util.Set;
 
-import static com.microsoft.intellij.ui.messages.AzureBundle.message;
-
-public class FunctionCliResolver {
-    public static String resolveFunc() throws IOException, InterruptedException {
-        final boolean isWindows = CommandUtils.isWindows();
-        final List<File> funCmdFiles = CommandUtils.resolvePathForCommand("func");
-        File result = null;
-        for (final File file : funCmdFiles) {
-            final File canonicalFile = file.getCanonicalFile();
-            if (!canonicalFile.exists()) {
-                continue;
-            }
-            // when `func core tools` is manually installed and func is available at PATH
-            // use canonical path to locate the real installation path
-            result = findFuncExecInFolder(canonicalFile.getParentFile(), isWindows);
-            if (result == null) {
-                if (isWindows) {
-                    result = resolveFuncForWindows(canonicalFile);
-                } else {
-                    // in linux/mac, when the way of `npm install azure-functions-core-tools`, the canonicalFile will point to `main.js`
-                    if (canonicalFile.getName().equals("main.js")) {
-                        result = findFuncExecInFolder(Paths.get(canonicalFile.getParent(), "..", "bin").normalize().toFile(),
-                                isWindows);
-                    }
+import static com.microsoft.azure.toolkit.lib.common.utils.CommandUtils.isWindows;
+
+public abstract class FunctionCliResolver {
+    public static String resolveFunc() {
+        return resolve().stream().findFirst().orElse(null);
+    }
+
+    @AzureOperation(name = "function.resolve_cli", type = AzureOperation.Type.TASK)
+    public static List<String> resolve() {
+        final FunctionCliResolver resolver = CommandUtils.isWindows() ? new WindowsFunctionCoreToolsResolver() : new LinuxFunctionCoreToolsResolver();
+        return resolver.resolveInner();
+    }
+
+    @Cacheable(cacheName = "function/resolve-cli")
+    private List<String> resolveInner() {
+        // resolve command from $PATH
+        final List<String> whichFuncDirs = resolveCommandPath("func");
+        final Set<String> results = new HashSet<>();
+        final Set<String> processedDirectories = new HashSet<>();
+        for (final String dir : whichFuncDirs) {
+            try {
+                final File canonicalFile = new File(dir).getCanonicalFile();
```

Comment:
what will happen if we move the `try catch` out of `for` ?

Reply:
no, for every dir in dirs, we need to skip the error and continue to search in dirs, if we move try out, a single error will break all the attempts